flask/helpers.py

```diff
@@ -190,7 +190,23 @@ def index():
     return current_app.make_response(args)
 
 
-def url_for(endpoint, **values):
+def url_for(*args, **kwargs):
+    """
+    wrapper based on https://github.com/miguelgrinberg/flack/blob/master/flack/utils.py
+    that also covers the case where application context exists with SERVER_NAME defined.
+    """
+    reqctx = _request_ctx_stack.top
+    appctx = _app_ctx_stack.top
+    if reqctx is None and (appctx is None or appctx.url_adapter is None):
+        if kwargs.get('_external', False):
+            raise RuntimeError('Cannot generate external URLs without a '
+                               'request context.')
+        with current_app.test_request_context():
```

Comment:
This feels extremely dirty. It's not the purpose of `url_for` to push a context.